### DISCUSSION TITLE
Difficulties creating a new pyodide package

### DISCUSSION THREAD

**Discussion by Jonah1234567** [pyodide/pyodide#3560] [Q&A]
I am trying to port python package with c++  for use in pyodide. The package is called simsopt https://github.com/hiddenSymmetries/simsopt and it is for optimizing stellarator reactor designs. After trying for a little while, I feel like I'm hitting a brick wall a little bit. I'm wondering if this port is possible without an unreasonable amount of effort?

Here is an error I get in the build process running with ubuntu 20.04

```
CMakeFiles/simsoptpp.dir/src/simsoptpp/python_distance.cpp.o
/usr/bin/c++ -DBOOST_ALL_NO_LIB -DFMT_LOCALE -Dsimsoptpp_EXPORTS -I/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/thirdparty/xtensor/include -I/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/thirdparty/xtensor-python/include -I/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/thirdparty/xsimd/include -I/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/thirdparty/xtl/include -I/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/thirdparty/eigen -I/tmp/build-env-464ldkj6/lib/python3.8/site-packages/numpy/core/include -I/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/src/simsoptpp -isystem /usr/include/python3.8 -O3 -march=native -mfma -ffp-contract=fast -O3 -DNDEBUG -fPIC -fvisibility=hidden -std=c++14 -flto -fno-fat-lto-objects -fvisibility=hidden -fopenmp -std=gnu++14 -MD -MT CMakeFiles/simsoptpp.dir/src/simsoptpp/python_distance.cpp.o -MF CMakeFiles/simsoptpp.dir/src/simsoptpp/python_distance.cpp.o.d -o CMakeFiles/simsoptpp.dir/src/simsoptpp/python_distance.cpp.o -c /tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/src/simsoptpp/python_distance.cpp
/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/src/simsoptpp/python_distance.cpp:172:24: error: variable or field ‘init_distance’ declared void
  172 | void init_distance(py::module_ &m){
      |                        ^~~~~~~
/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/src/simsoptpp/python_distance.cpp:172:24: error: ‘module_’ is not a member of ‘py’; did you mean ‘module’?
  172 | void init_distance(py::module_ &m){
      |                        ^~~~~~~
      |                        module
/tmp/build-via-sdist-1skygobz/simsopt-0.12.2.post6+g8db94519.d20230209.dirty/src/simsoptpp/python_distance.cpp:172:33: error: ‘m’ was not declared in this scope; did you mean ‘tm’?
  172 | void init_distance(py::module_ &m){
      |                                 ^
      |                                 tm
```

I feel like this error suggests that the C++ for this project is written in way that makes it near impossible to port. Please let me know if you think I should keep trying or just give up on it or if there is something I am missing.

**Comment by hoodmane** [comment]
That file looks like a good candidate for porting. Should be fine if configured correctly. What did you do to try to build it?

**Comment by Jonah1234567** [comment]
I ran through the out of tree build process outlined in the docs. Then there were some issues with the dependencies that I had to fix. Could it be a pybind issue or something like that?

**Reply by hoodmane** [reply]
I am not sure. I think CMake packages are still a bit experimental. @ryanking13 knows more about CMake so perhaps he can provide some suggestions.

**Reply by ryanking13** [reply]
What happens if you build the package natively? with following command:

```
pip install build
python -m build .
```

**Reply by Jonah1234567** [reply]
The build works fine, although it takes some tinkering with dependencies, they are not downloaded properly with the package. A whl file is generated at the end

**Reply by ryanking13** [reply]
Okay, thanks for the check. We also had a pybind11 issue in duckdb before (which is not resolved yet) https://github.com/pyodide/pyodide/pull/2961, so maybe there are some issue of using pybind11 in Pyodide build system... I'll try to reproduce the issue when I have bandwidth.

**Reply by ryanking13** [reply]
Acutally, I am failing earlier with following CMake error when I tried to reproduce.

```
-- Boost include dirs are /src/packages/simsopt/build/simsopt-0.12.2/build/temp.emscripten_3_1_31_wasm32-cpython-311/thirdparty/boost/src/boost
-- pybind11 v2.6.2 dev1
-- Found PythonInterp: /tmp/build-env-6e_705o_/bin/python (found version "3.11.1") 
CMake Error at thirdparty/pybind11/tools/FindPythonLibsNew.cmake:161 (message):
  Python config failure: Python is 64-bit, chosen compiler is 32-bit
Call Stack (most recent call first):
  thirdparty/pybind11/tools/pybind11Tools.cmake:38 (find_package)
  thirdparty/pybind11/tools/pybind11Common.cmake:201 (include)
  thirdparty/pybind11/CMakeLists.txt:169 (include)
```

**Reply by Jonah1234567** [reply]
I haven't seen that one, maybe double check the that simsopt's pybind11 under thirdparty is being downloaded, I had some issues with that earlier. I made some progress and now my main issue is that OpenMP isn't being found but I think that's an issue specific to my machine.

**Reply by Jonah1234567** [reply]
Never mind the above, I just got that error

**Reply by hoodmane** [reply]
> OpenMP

We use single-threaded Emscripten, so threading / multiprocessing libraries may cause problems. Is there a flag to make a single threaded build?

**Reply by Jonah1234567** [reply]
From a very quick glance at the repo it doesn't appear there is a threading flag, but I am going to look further. I tried running a slightly different pyodide build with new flags

pyodide build cmake -DCMAKE_C_COMPILER=gcc-64 -DCMAKE_CXX_COMPILER=g++-64`

this yielded different results:
`-- Checking if compiler supports needed C++17 constructs
-- Checking if compiler supports needed C++17 constructs - no
-- Checking if compiler supports needed C++14 constructs
-- Checking if compiler supports needed C++14 constructs - no
-- Checking if compiler supports C++ make_unique
-- Checking if compiler supports C++ make_unique - no
-- Checking if compiler supports C++ unique_ptr
-- Checking if compiler supports C++ unique_ptr - no
CMake Error at CMakeLists.txt:101 (message):
  The C++ compiler does not support C++11 (e.g.  std::unique_ptr).`

**Reply by ryanking13** [reply]
`-DCMAKE_C_COMPILER=gcc-64 -DCMAKE_CXX_COMPILER=g++-64`

Please don't override compilers. Overriding the compiler toolchain will break the build, since we use our own compiler wrappers.

**Reply by ryanking13** [reply]
It seems like our [CMake toolchain file](https://github.com/pyodide/pyodide/blob/main/pyodide-build/pyodide_build/tools/cmake/Modules/Platform/Emscripten.cmake) is not applied when doing out-of-tree build, which is a bug.

**Reply by Jonah1234567** [reply]
Is there a fix for that I can work on or should I just open up an issue?

**Reply by ryanking13** [reply]
I think it will take some time for me to figure out what is happening exactly. Please open an issue with some commands you tried.

**Reply by Jonah1234567** [reply]
issue is opened
